File: navi/src/main/java/org/kexie/android/dng/navi/widget/CarMarker.java
package org.kexie.android.dng.navi.widget;

import android.animation.PropertyValuesHolder;
import android.animation.TypeEvaluator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.view.animation.LinearInterpolator;

import com.amap.api.maps.AMap;
import com.amap.api.maps.CameraUpdateFactory;
import com.amap.api.maps.model.BitmapDescriptor;
import com.amap.api.maps.model.BitmapDescriptorFactory;
import com.amap.api.maps.model.CameraPosition;
import com.amap.api.maps.model.LatLng;
import com.amap.api.maps.model.Marker;
import com.amap.api.maps.model.MarkerOptions;

import org.kexie.android.dng.navi.R;
import org.kexie.android.dng.navi.model.beans.Point;



@SuppressWarnings("WeakerAccess")
public class CarMarker {

    private final AMap map;
    private final Marker car;
    private final Marker direction;

    private long time = BASE_TIME;
    private float zoom = 20f;
    private boolean isLock;
    private Pos last;
    private ValueAnimator animator;

    private static final TypeEvaluator<Pos> POS_TYPE_EVALUATOR
            = (fraction, startValue, endValue) -> {
        double x = endValue.point.getLongitude() - startValue.point.getLongitude();
        double y = endValue.point.getLatitude() - startValue.point.getLatitude();
        Point point = Point.form(startValue.point.getLongitude() + x * fraction,
                startValue.point.getLatitude() + y * fraction);
        float r = endValue.bearing - startValue.bearing;
        if (r > 180) {
            r -= 360;
        } else if (r < -180) {
            r += 360;
        }
        float bearing = (startValue.bearing + r * fraction);
        return new Pos(point, bearing);
    };
    private static final int BASE_TIME = 50;
    private static final float TILT = 80f;

    public boolean isLock() {
        return isLock;
    }

    public void setLock(boolean lock) {
        isLock = lock;
    }

    public void setZoom(float zoom) {
        this.zoom = zoom;
    }

    public boolean isVisible() {
        return car.isVisible() && direction.isVisible();
    }

    public void setVisible(boolean value) {
        car.setVisible(value);
        direction.setVisible(value);
        car.setToTop();
        direction.setToTop();
    }

    public CarMarker(AMap aMap, Bitmap car, Bitmap direction) {
        this.map = aMap;
        this.car = getMarker(aMap, car);
        this.direction = getMarker(aMap, direction);
        setVisible(false);
    }

    public static CarMarker getDefault(Context context, AMap aMap) {
        Resources resources = context.getResources();
        Bitmap car = BitmapFactory.decodeResource(resources, R.drawable.caricon);
        Bitmap direction = BitmapFactory.decodeResource(resources, R.drawable.navi_direction);
        return new CarMarker(aMap, car, direction);
    }

    public void draw(Point location, float bearing) {
        ValueAnimator animator = this.animator;
        if (animator == null) {
            last = new Pos(location, bearing);
            apply(last);
            ValueAnimator newAnimator = new ValueAnimator();
            newAnimator.setInterpolator(new LinearInterpolator());
            newAnimator.setDuration(time);
            newAnimator.addUpdateListener(animation -> {
                last = (Pos) animation.getAnimatedValue();
                apply(last);
            });
            this.animator = newAnimator;
        } else {
            if (animator.isRunning()) {
                animator.cancel();
                if (time > BASE_TIME) {
                    time /= 2;
                } else {
                    time *= 2;
                }
            }
            animator.setValues(PropertyValuesHolder.ofObject(
                    "",
                    POS_TYPE_EVALUATOR,
                    last,
                    new Pos(location, bearing)));
            animator.setDuration(time);
            animator.start();
            this.animator = animator;
        }
    }

    private static Marker getMarker(AMap map, Bitmap bitmap) {
        BitmapDescriptor descriptor = BitmapDescriptorFactory
                .fromBitmap(bitmap);
        return map.addMarker(new MarkerOptions()
                .anchor(0.5f, 0.5f)
                .position(new LatLng(0.0, 0.0))
                .setFlat(true)
                .icon(descriptor));
    }

    private void apply(Pos it) {
        LatLng point = it.point.unBox(LatLng.class);
        direction.setPosition(point);
        car.setPosition(point);
        car.setRotateAngle(360 - it.bearing);
        setVisible(true);
        if (!isLock) {
            return;
        }
        map.moveCamera(CameraUpdateFactory.newCameraPosition(
                CameraPosition.builder()
                        .tilt(TILT)
                        .zoom(zoom)
                        .bearing(it.bearing)
                        .target(point)
                        .build()));
    }

    private static final class Pos {
        final Point point;
        final float bearing;

        private Pos(Point point, float bearing) {
            this.point = point;
            this.bearing = bearing;
        }
    }
}

File: navi/src/main/java/org/kexie/android/dng/navi/widget/CarMarker.kt
package org.kexie.android.dng.navi.widget

import android.animation.PropertyValuesHolder
import android.animation.TypeEvaluator
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.BitmapFactory
import android.view.animation.LinearInterpolator
import com.amap.api.maps.AMap
import com.amap.api.maps.CameraUpdateFactory
import com.amap.api.maps.model.*
import org.kexie.android.dng.navi.R
import org.kexie.android.dng.navi.model.Point

class CarMarker(context: Context, private val map: AMap) {

    private val context = context.applicationContext;
    private val carMarker: Marker
    private var directionMarker: Marker
    private var valueAnimator: ValueAnimator? = null
    private lateinit var last: Location
    private var time = BASE_TIME
    var isVisible: Boolean
        get() = carMarker.isVisible && directionMarker.isVisible
        set(value) {
            carMarker.isVisible = value;
            directionMarker.isVisible = value
            directionMarker.setToTop()
            carMarker.setToTop()
        }
    var zoom: Float = 20f
    var isLock: Boolean = false

    private data class Location(var point: Point, var bearing: Float)

    init {
        fun getMarker(res: Int): Marker {
            val descriptor = BitmapDescriptorFactory
                    .fromBitmap(BitmapFactory.decodeResource(this.context
                            .resources, res))
            return map.addMarker(MarkerOptions()
                    .anchor(0.5f, 0.5f)
                    .position(LatLng(0.0, 0.0))
                    .setFlat(true)
                    .icon(descriptor))
        }
        carMarker = getMarker(R.drawable.caricon)
        directionMarker = getMarker(R.drawable.navi_direction)
        isVisible = false
    }

    fun draw(location: Point, bearing: Float) {
        fun apply(it: Location) {
            val point = it.point.unBox(LatLng::class.java)
            directionMarker.position = point
            carMarker.position = point
            carMarker.rotateAngle = 360 - it.bearing
            isVisible = true
            if (!isLock) {
                return
            }
            map.moveCamera(CameraUpdateFactory.newCameraPosition(CameraPosition.Builder()
                    .tilt(TILT)
                    .zoom(zoom)
                    .bearing(it.bearing)
                    .target(point)
                    .build()))
        }
        val animator = valueAnimator;
        if (animator == null) {
            last = Location(location, bearing)
            apply(last)
            val newAnimator = ValueAnimator()
            newAnimator.interpolator = interpolator
            newAnimator.duration = time
            newAnimator.addUpdateListener { animation ->
                last = animation.animatedValue as Location
                apply(last)
            }
            valueAnimator = newAnimator
        } else {
            if (animator.isRunning) {
                animator.cancel()
                if (time > BASE_TIME) {
                    time /= 2
                } else {
                    time *= 2
                }
            }
            animator.setValues(PropertyValuesHolder.ofObject("",
                    evaluator,
                    last,
                    Location(location, bearing)))
            animator.duration = time
            animator.start()
            valueAnimator = animator
        }
    }

    companion object {
        private val interpolator = LinearInterpolator()
        private val evaluator = object : TypeEvaluator<Location> {
            override fun evaluate(fraction: Float, startValue: Location, endValue: Location): Location {
                val x = endValue.point.longitude - startValue.point.longitude
                val y = endValue.point.latitude - startValue.point.latitude
                val point = Point.form(startValue.point.longitude + x * fraction,
                        startValue.point.latitude + y * fraction);
                var r = endValue.bearing - startValue.bearing
                if (r > 180) {
                    r -= 360
                } else if (r < -180) {
                    r += 360
                }
                val bearing = (startValue.bearing + r * fraction)
                return Location(point, bearing)
            }
        }
        private const val BASE_TIME: Long = 50
        private const val TILT = 80f
    }
}